### PR TITLE
mTime values will be readout from within nc-file (seconds of day), not from filename

### PR DESCRIPTION
**Pull request by ulysses78**
in matlabfile ./lib/io/readPollyRawData.m
mTime values will now be readout from within nc-file (seconds of day), not from filename any longer.
When merging level0-nc-files to one 24h file, it is  usefull to read the seconds of day from within the nc-file.

### CODE REVIEW

**Comment by ulysses78** [comment]
@ZPYin 
Why have you choosen to readout the measurement_time (mTime) from filename, not from the seconds per day from within the level0_nc-file? Sometimes I have seen very small files (<100000 Byte), which contain wrong dates and seconds (19700101, seconds 0). But those files get filtered anyway because of their size. Any other reason?

File: lib/io/readPollyRawData.m [hunk 1]
```diff
@@ -154,6 +154,8 @@
     mTime(2, :) = thisHour * 3600 + ...
                  thisMinute * 60 + ...
                  thisSecond + 30 .* (0:(size(mTime, 2) - 1));
+
+    mTime = ncread(file, 'measurement_time');
```
Comment (ZPYin) [inline lib/io/readPollyRawData.m:158]: `mTime` was in fact read from variable `measurement_time` within the nc file, please see line 101 of `readPollyRawData`.

So please confirm whether this fix is necessary.

**Review by ZPYin — COMMENTED** [review]
Please see my in-line comment.

**Comment by ZPYin** [comment]
> @ZPYin Why have you choosen to readout the measurement_time (mTime) from filename, not from the seconds per day from within the level0_nc-file? Sometimes I have seen very small files (<100000 Byte), which contain wrong dates and seconds (19700101, seconds 0). But those files get filtered anyway because of their size. Any other reason?

I need to add some background into this topic. 

Holger (@HolgerPollyNet ) mentioned that in some **old** polly system, the data recording system may have issue of stopping each profile accurately after 30 seconds, or very **strange timestamps** as you mentioned. Therefore, Holger suggests to implement an additional way to correct the timestamps according to the start time and fixed temporal interval. Holger told me that the **time registered in the filename is always correct**, which can serve as the basis to recover timestamps for each signal profile, see below (you can use polly_global_config keywrod to turn on this feature)

https://github.com/PollyNET/Pollynet_Processing_Chain/blob/48a9864703a3f6dba666648d10e40234e8d84468/lib/preprocess/pollyPreprocess.m#L255-L259

So that's why you see I usually chose to read start time from the filename.

Regarding data filtering, yes, we used the data filesize as a threshold to exclude empty files or small data file without enough measurements, see below

https://github.com/PollyNET/Pollynet_Processing_Chain/blob/48a9864703a3f6dba666648d10e40234e8d84468/lib/io/writeTodoListAuto.m#L165-L168

File: lib/io/readPollyRawData.m (hunk 1 above)
Comment (ulysses78) [reply lib/io/readPollyRawData.m:158]: Yes, in line 101 you define the mTime by reading within the nc-file, but in line 153 & 154 you redefine it by reading the filename.

**Comment by ulysses78** [comment]
> > @ZPYin Why have you choosen to readout the measurement_time (mTime) from filename, not from the seconds per day from within the level0_nc-file? Sometimes I have seen very small files (<100000 Byte), which contain wrong dates and seconds (19700101, seconds 0). But those files get filtered anyway because of their size. Any other reason?
> 
> I need to add some background into this topic.
> 
> Holger (@HolgerPollyNet ) mentioned that in some **old** polly system, the data recording system may have issue of stopping each profile accurately after 30 seconds, or very **strange timestamps** as you mentioned. Therefore, Holger suggests to implement an additional way to correct the timestamps according to the start time and fixed temporal interval. Holger told me that the **time registered in the filename is always correct**, which can serve as the basis to recover timestamps for each signal profile, see below (you can use polly_global_config keywrod to turn on this feature)
> 
> https://github.com/PollyNET/Pollynet_Processing_Chain/blob/48a9864703a3f6dba666648d10e40234e8d84468/lib/preprocess/pollyPreprocess.m#L255-L259
> 
> So that's why you see I usually chose to read start time from the filename.
> 
> Regarding data filtering, yes, we used the data filesize as a threshold to exclude empty files or small data file without enough measurements, see below
> 
> https://github.com/PollyNET/Pollynet_Processing_Chain/blob/48a9864703a3f6dba666648d10e40234e8d84468/lib/io/writeTodoListAuto.m#L165-L168

OK. I understand. My goal was to merge the level0 files from one day (24h) and process it with picasso. Finally I would get some nice 24h-plots. But for days where the measurement has some gaps the merged file will have continuous timestamps starting from the filenames timestamp - no gaps will be recognized.
So I guess the best would be to check the timestamps in the nc-file. If they are ok, then take them, otherwise choose the timestamps from filename. It would be interesting to know which polly-systems show those timestamp errors.
Another way for the solution of my problem would be to merge all the nc-files after they have been processed with picasso.

**Comment by ZPYin** [comment]
> So I guess the best would be to check the timestamps in the nc-file. If they are ok, then take them, otherwise choose the timestamps from filename. ...Another way for the solution of my problem would be to merge all the nc-files after they have been processed with picasso.

What about filling in the merged file with NaN values to represent the gaps? Picasso can treat such profiles as invalid measurements, and render these gaps as black strips in the plots. 

But you may need to consider the loss of accurate measurement time either in the lev0 data file mergeing or after the data processing with Picasso. Because in either way, the continuous measurement will be re-binned to a fixed temporal grid. Then the tiny shift of measurement time (< 30 s) will disappear.

> It would be interesting to know which polly-systems show those timestamp errors.

This is the issue for pollyxt_tropos once in Haifa (wrong timestamp) and polly_1st (inaccurate 30 second interval). 
I don't remember too much details. I have sent a short report to Holger through TROPOS email. I need to check it in my old laptop.

**Comment by ulysses78** [comment]
On friday or next week I will talk to Holger about this issue. He can tell me what solution he prefers ;-)

**Comment by HolgerPollyNet** [comment]
Hi, I would prefer to keep the old options for those files for which the DAQ had a wrong date (mostly 1970), because the data itself is useful. But, maybe one has to process these rare data manually using the forcetime flag. For the operational procedure one can of course use the time from the netcdf. This also alerts us, in case the DAQ has a wrong time again.

**Pull request closed without merging**